### PR TITLE
[oneDPL][hetero] A blocking behavior fix for some algo patterns (sync API)

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
[oneDPL][hetero] + a blocking behavior fix for some algo patterns (sync API)

According to 4.7.2.3. Buffer synchronization rules, rule number (1), a destructor of such buffer does not need to block. So, RAII instance of such buffer doesn't guarantee implicit wait. The PR fixes the behavior for oneDPL sync API.

(https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:buf-sync-rules)

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1034]: I don't think any comment is required here anymore.  If we want some comment here, perhaps: 

//__pattern_walk2 is a blocking call

Similar comment for the repetitions of this later.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1667]: I dont think a comment is necessary here.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1665]: I believe I agree with the clang-format suggestion to move `.wait()` to the next line, but it is your call.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1328]: Do we have any guarantee that the first `__pattern_walk2()` call above returns?
I think these are independent from each other, and were both depending on the destructors for synchronization (or so we thought...).

I think it may need to be made synchronous as well.  It would be nice to have it return an event which could be waited on so that these functions could run concurrently, but that is likely beyond the scope of this PR.

**Review by SergeyKopienko — DISMISSED** [review]
LGTM

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1328]: 1) These patterns (pattern_merge and pattern_walk2) have  an implicit dependence via SYCL dependency graph.
2) In other pattern (__pattern_stable_partition) we have dependent working patterns, and those can be launched asynchrony and we have to wait two events by `wait(e1, e2)`. There is a comment about that in the code.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1034]: I think to keep such comment is very important , because the situation (with misunderstanding what is going on)  might be repeated again by looking at the someone? not me, you or Sergey.. And that person also may decide that __pattern_walk2 may be called in async way "because" we have sycl::buffer above on the stack.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1034]: Fair point, perhaps we just need to expand it a little more then:
```
// The temporary buffer is constructed from a range, therefore it's destructor will not block, therefore 
// we must call __pattern_walk2 in a way which provides blocking synchronization for this pattern.
```

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1328]: For `__pattern_stable_partition`, line 1325, we call `__pattern_walk2` without synchronization.  It has no dependencies with the blocking call on line 1331 as far as I can see.  

Therefore, we have no guarantee that the call from line 1325 completes before we return from `__pattern_stable_partition`.  Ideally `__pattern_walk2` would return a future or event we could wait on, but it does not, so I think we just need to call it in a blocking fashion, unless we want more refactoring.

I'm missing where you mention we wait on 2 events as you describe: `wait(e1, e2)`.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1667]: A similar adjustment to the comment as mentioned above would work instead.

File: include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h:1331]: yes, I agree with that.

**Review by danhoeflinger — DISMISSED** [review]
I'd recommend fixing the clang-format issues (there are some for comments and others for code).  Otherwise this LGTM.

Also lets wait for green CI.

**Comment by MikeDvorskiy** [comment]
> I'd recommend fixing the clang-format issues (there are some for comments and others for code). Otherwise this LGTM.
> 
> Also lets wait for green CI.

I've just applied of the most of clang format recommendations, excepting one case with ".wait()"

**Review by danhoeflinger — APPROVED** [review]
LGTM, wait for green CI to merge please.

**Pull request merged**
